### ISSUE TITLE
Multiple errors during installation on CentOS 7

### ISSUE THREAD

**Issue by verdurin** [gc3-uzh-ch/elasticluster#268]
Trying a new installation on a CentOS 7.2 VM from `master` and I see the following error:

```
Installed /home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/pyasn1-0.1.9-py2.7.egg
Searching for cryptography>=1.1
Reading https://pypi.python.org/simple/cryptography/
Best match: cryptography 1.4
Downloading https://pypi.python.org/packages/a9/5b/a383b3a778609fe8177bd51307b5ebeee369b353550675353f46cb99c6f0/cryptography-1.4.tar.gz#md5=a9763e3831cc7cdb402c028fac1ceb39
Processing cryptography-1.4.tar.gz
Writing /tmp/easy_install-3rnLJG/cryptography-1.4/setup.cfg
Running cryptography-1.4/setup.py -q bdist_egg --dist-dir /tmp/easy_install-3rnLJG/cryptography-1.4/egg-dist-tmp-bNF2C4
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libffi' found
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libffi' found
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libffi' found
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libffi' found
Package libffi was not found in the pkg-config search path.
Perhaps you should add the directory containing `libffi.pc'
to the PKG_CONFIG_PATH environment variable
No package 'libffi' found
c/_cffi_backend.c:15:17: fatal error: ffi.h: No such file or directory
 #include <ffi.h>
                 ^
compilation terminated.
error: Setup script exited with error: command 'gcc' failed with exit status 1
```

**Comment by verdurin** [comment]
The following RPM is already installed on the system:

`python2-cffi-1.5.2-1.el7.x86_64`

**Comment by verdurin** [comment]
Installing `libffi-devel` fixed this.

**Comment by verdurin** [comment]
Next error:

```
Searching for debtcollector>=1.2.0
Reading https://pypi.python.org/simple/debtcollector/
Best match: debtcollector 1.4.0
Downloading https://pypi.python.org/packages/19/f2/a2cf6e62bcfa4c3428bdcba09150e5f6835db3797d74689d4427510a7532/debtcollector-1.4.0.tar.gz#md5=46c22a03f5a10abf60b594b819d31779
Processing debtcollector-1.4.0.tar.gz
Writing /tmp/easy_install-f5Gdnw/debtcollector-1.4.0/setup.cfg
Running debtcollector-1.4.0/setup.py -q bdist_egg --dist-dir /tmp/easy_install-f5Gdnw/debtcollector-1.4.0/egg-dist-tmp-zBh94n
ERROR:root:Error parsing
Traceback (most recent call last):
  File "/home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/pbr-1.10.0-py2.7.egg/pbr/core.py", line 111, in pbr
    attrs = util.cfg_to_args(path, dist.script_args)
  File "/home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/pbr-1.10.0-py2.7.egg/pbr/util.py", line 264, in cfg_to_args
    wrap_commands(kwargs)
  File "/home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/pbr-1.10.0-py2.7.egg/pbr/util.py", line 566, in wrap_commands
    cmdclass = ep.resolve()
  File "build/bdist.linux-x86_64/egg/pkg_resources/__init__.py", line 2216, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/Babel-2.3.4-py2.7.egg/babel/messages/__init__.py", line 12, in <module>
    from babel.messages.catalog import *
  File "/home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/Babel-2.3.4-py2.7.egg/babel/messages/catalog.py", line 23, in <module>
    from babel.dates import format_datetime
  File "/home/centos/src/venv-2016-06-07/elasticluster/lib/python2.7/site-packages/Babel-2.3.4-py2.7.egg/babel/dates.py", line 23, in <module>
    import pytz as _pytz
ImportError: No module named pytz
error: Setup script exited with error in setup command: Error parsing /tmp/easy_install-f5Gdnw/debtcollector-1.4.0/setup.cfg: ImportError: No module named pytz
```

**Comment by riccardomurri** [comment]
> Trying a new installation on a CentOS 7.2 VM from `master` and I see the following error:
>
> ```
> c/_cffi_backend.c:15:17: fatal error: ffi.h: No such file or directory
> ```

Thanks for reporting and suggesting a solution!

I have updated the "troubleshooting" section accordingly.

As far as I understand, this is the best I can do now as Python's
`setup.py` gives me no control or way to check what system packages are
installed.

Furthermore, `cffi` is a dependency three levels down the dependency
tree (elasticluster -> paramiko -> cryptography -> cffi) so we really
have no control over it...

**Comment by riccardomurri** [comment]
I think the bug does not lie actually in ElastiCluster's code: package
`debtcollector` (a dependency of the `novaclient` package used by
ElastiCluster) uses `pbr` instead of the plain `setup.py`.  Then
something happens, and `pbr` wants to print a message, and tries to use
[Babel][1] to translate the message.  Now `Babel` requires `pytz` to run
(and duly lists it as a dependency), but since neither
[debtcollector][2] nor `pbr` list `Babel` as a requirement, it has not
been properly installed: in particular, its dependency `pytz` has not
yet been unpacked...

Can you please try with branch "issues/#268" from
http://github.com/riccardomurri/elasticluster and see if it works around
the issue?


[1]: https://pypi.python.org/pypi/Babel/2.3.4
[2]: https://github.com/openstack/debtcollector/commit/c204667ed8eec848b664a41a8e5869771c5dd9a4

**Comment by verdurin** [comment]
@riccardomurri - yes, certainly not an Elasticluster bug per se.

I've tried with that branch and I'm now able to install normally. It does take a bit longer than normal, with these extra dependencies, but that's fine.

Thanks.

**Comment by riccardomurri** [comment]
Finally merged into "master" in commit fa4b8aa
